### ISSUE TITLE
Koreader OPDS and Wireless Calibre function keep asking for wifi prompt in WSL2 Linux

### ISSUE THREAD

**Issue by zephyros-dev** [koreader/koreader#10694]
* KOReader version: v2023.06.1
* Device: Windows 10 19045.3208, WSL2, Debian GNU/Linux 12 (bookworm)

#### Issue
I cannot connect to OPDS. The popup "Turn on Wifi" shows up for connection attempt. The PC is connected via Ethernet. Ping can be made normally inside WSL VM

#### Steps to reproduce
1. Install debian via wsl in a powershell
```
wsl --install -d Debian
```
2. Install koreader via .deb
```
sudo apt update
sudo apt install curl
curl -Lo koreader.deb https://github.com/koreader/koreader/releases/download/v2023.06.1/koreader-2023.06.1-amd64.deb
sudo apt install -f ./koreader.deb
```
4. Add ping capabilities for debian via the following command, or do this https://github.com/microsoft/WSL/issues/5109#issuecomment-939676832
```
sudo setcap cap_net_raw+p /bin/ping
```
5. Run koreader
```
koreader -v
```
6. Go to Search > OPDS Catalogs > Select a configured OPDS Catalog
7. A popup asking if we should turn on wifi is shown
8. On the next attempt the popup is shown again

##### `crash.log` (if applicable)
<details>
  <summary>`ip r` output</summary>

```
default via 172.27.96.1 dev eth0 proto kernel 
172.27.96.0/20 dev eth0 proto kernel scope link src 172.27.107.6
```

</details>

<details>
  <summary>`ip a` output</summary>

```
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 00:15:5d:d1:3a:29 brd ff:ff:ff:ff:ff:ff
    inet 172.27.107.6/20 brd 172.27.111.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::215:5dff:fed1:3a29/64 scope link
       valid_lft forever preferred_lft forever
```
</details>

<details>
  <summary>`koreader -v` output</summary>

```

---------------------------------------------
                launching...
  _  _____  ____                _
 | |/ / _ \|  _ \ ___  __ _  __| | ___ _ __
 | ' / | | | |_) / _ \/ _` |/ _` |/ _ \ '__|
 | . \ |_| |  _ <  __/ (_| | (_| |  __/ |
 |_|\_\___/|_| \_\___|\__,_|\__,_|\___|_|

 It's a scroll... It's a codex... It's KOReader!

 [*] Current time: 07/14/23-21:33:33
 [*] Version: v2023.06.1

ffi.load: libs/libutf8proc.so.2
ffi.load: blitbuffer
ffi.load: SDL2
ffi.load (warning): libSDL2.so: cannot open shared object file: No such file or directory
ffi.load: libSDL2-2.0.so
ffi.load (warning): libSDL2-2.0.so: cannot open shared object file: No such file or directory
ffi.load: libSDL2-2.0.so.0
ffi.load: SDL2
ffi.load (warning): libSDL2.so: cannot open shared object file: No such file or directory
ffi.load: libSDL2-2.0.so
ffi.load (warning): libSDL2-2.0.so: cannot open shared object file: No such file or directory
ffi.load: libSDL2-2.0.so.0
Starting SDL in /usr/lib/koreader/
ffi.load: libs/liblodepng.so
07/14/23-21:33:33 INFO  initializing for device Linux 
07/14/23-21:33:33 INFO  framebuffer resolution: {
  h = 1887,
  w = 1068
} --[[table: 0x7f3d4770bb60]]
ffi.load: libs/libmupdf.so
ffi.load: libs/libwrap-mupdf.so
ffi.load: libs/libfreetype.so.6
ffi.load: libs/libharfbuzz.so.0
ffi.load: libs/libzstd.so.1
07/14/23-21:33:35 INFO  Loading plugins from directory: plugins 
ffi.load: sqlite3
07/14/23-21:33:36 INFO  Loading plugins from directory: /home/user/.config/koreader/plugins/ 
07/14/23-21:33:36 DEBUG triggering refresh {
  mode = "ui",
  region = "1068x1887+0+0"
} --[[table: 0x7f3d420e1a70]]
07/14/23-21:33:37 DEBUG triggering refresh {
  mode = "flashui",
  region = "1068x772+0+0"
} --[[table: 0x7f3d41c28788]]
07/14/23-21:33:43 DEBUG triggering refresh {
  mode = "fast",
  region = "93x51+288+9"
} --[[table: 0x7f3d418d3800]]
07/14/23-21:33:43 DEBUG triggering refresh {
  mode = "ui",
  region = "1068x772+0+0"
} --[[table: 0x7f3d41899080]]
07/14/23-21:33:45 DEBUG triggering refresh {
  mode = "fast",
  region = "1054x59+7+589"
} --[[table: 0x7f3d4186c9b8]]
07/14/23-21:33:45 DEBUG triggering refresh {
  mode = "flashui",
  region = "1068x1887+0+0"
} --[[table: 0x7f3d4319c3f0]]
07/14/23-21:33:47 DEBUG triggering refresh {
  mode = "fast",
  region = "1068x127+0+313"
} --[[table: 0x7f3d41c71c90]]
07/14/23-21:33:49 DEBUG triggering refresh {
  mode = "ui",
  region = "1068x127+0+313"
} --[[table: 0x7f3d431a1df8]]
07/14/23-21:33:49 DEBUG triggering refresh {
  mode = "ui",
  region = "808x152+130+867"
} --[[table: 0x7f3d424d04f0]]
07/14/23-21:33:50 DEBUG triggering refresh {
  mode = "fast",
  region = "387x54+535+953"
} --[[table: 0x7f3d424bf068]]
07/14/23-21:33:50 DEBUG triggering refresh {
  mode = "ui",
  region = "794x98+137+909"
} --[[table: 0x7f3d41c32770]]
07/14/23-21:33:50 DEBUG triggering refresh {
  mode = "ui",
  region = "808x152+130+867"
} --[[table: 0x7f3d411e5588]]

```

</details>

**Comment by NiLuJe** [comment]
c.f., https://github.com/koreader/koreader/issues/10087#issuecomment-1636122720

Very likely fixed by #10669, please try on master.

**Comment by Frenzie** [comment]
Here's the latest Debian build: https://gitlab.com/koreader/nightly-builds/-/jobs/4652554079

**Comment by zephyros-dev** [comment]
> Here's the latest Debian build: https://gitlab.com/koreader/nightly-builds/-/jobs/4652554079
The nightly build does not seems to help unfortunately. 
Regarding the SDL, the WSL Debian don't have them by default, so after I setup them using the following commands:
```
sudo apt install libsdl2-dev libsdl2-2.0-0 -y;
sudo apt install libfreetype6-dev libsdl2-ttf-dev libsdl2-ttf-2.0-0 -y;
```

The logs doesn't seems to complain about SDL anymore

<details>
  <summary>`koreader -v` output with SDL installed</summary>

```
---------------------------------------------
                launching...
  _  _____  ____                _
 | |/ / _ \|  _ \ ___  __ _  __| | ___ _ __
 | ' / | | | |_) / _ \/ _` |/ _` |/ _ \ '__|
 | . \ |_| |  _ <  __/ (_| | (_| |  __/ |
 |_|\_\___/|_| \_\___|\__,_|\__,_|\___|_|

 It's a scroll... It's a codex... It's KOReader!

 [*] Current time: 07/15/23-10:25:52
 [*] Version: v2023.06.1-22-g3e43a21c_2023-07-14

ffi.load: libs/libutf8proc.so.2
ffi.load: blitbuffer
ffi.load: SDL2
ffi.load: SDL2
Starting SDL in /usr/lib/koreader/
ffi.load: libs/liblodepng.so
07/15/23-10:25:52 INFO  initializing for device Linux 
07/15/23-10:25:52 INFO  framebuffer resolution: {
  h = 1887,
  w = 1068
} --[[table: 0x7f4202413830]]
ffi.load: libs/libmupdf.so
ffi.load: libs/libwrap-mupdf.so
ffi.load: libs/libfreetype.so.6
ffi.load: libs/libharfbuzz.so.0
ffi.load: libs/libzstd.so.1
07/15/23-10:25:54 INFO  Loading plugins from directory: plugins 
ffi.load: sqlite3
07/15/23-10:25:54 INFO  Loading plugins from directory: /home/user/.config/koreader/plugins/ 
07/15/23-10:25:54 DEBUG triggering refresh {
  mode = "ui",
  region = "1068x1887+0+0"
} --[[table: 0x7f41c41c0420]]
07/15/23-10:25:57 DEBUG triggering refresh {
  mode = "flashui",
  region = "1068x716+0+0"
} --[[table: 0x7f41c3847020]]
07/15/23-10:25:58 DEBUG triggering refresh {
  mode = "fast",
  region = "1054x59+7+69"
} --[[table: 0x7f41c3c15498]]
07/15/23-10:25:58 DEBUG triggering refresh {
  mode = "ui",
  region = "1068x716+0+0"
} --[[table: 0x7f41c6db7810]]
07/15/23-10:25:59 DEBUG triggering refresh {
  mode = "fast",
  region = "93x51+192+9"
} --[[table: 0x7f41c6dd1538]]
07/15/23-10:25:59 DEBUG triggering refresh {
  mode = "ui",
  region = "1068x772+0+0"
} --[[table: 0x7f41c6e20f58]]
07/15/23-10:26:00 DEBUG triggering refresh {
  mode = "fast",
  region = "93x51+288+9"
} --[[table: 0x7f41c6e2d088]]
07/15/23-10:26:00 DEBUG triggering refresh {
  mode = "ui",
  region = "1068x772+0+0"
} --[[table: 0x7f41c6e75ba0]]
07/15/23-10:26:01 DEBUG triggering refresh {
  mode = "fast",
  region = "1054x59+7+589"
} --[[table: 0x7f41c6e6e300]]
07/15/23-10:26:01 DEBUG triggering refresh {
  mode = "flashui",
  region = "1068x1887+0+0"
} --[[table: 0x7f41c3c13b90]]
07/15/23-10:26:01 DEBUG triggering refresh {
  mode = "fast",
  region = "1068x127+0+563"
} --[[table: 0x7f41c40d2588]]
07/15/23-10:26:03 DEBUG triggering refresh {
  mode = "ui",
  region = "1068x127+0+563"
} --[[table: 0x7f41c402c5f8]]
07/15/23-10:26:03 DEBUG triggering refresh {
  mode = "ui",
  region = "808x152+130+867"
} --[[table: 0x7f41c40cb560]]
07/15/23-10:26:05 DEBUG triggering refresh {
  mode = "fast",
  region = "387x54+535+953"
} --[[table: 0x7f41c6d8d658]]
07/15/23-10:26:05 DEBUG triggering refresh {
  mode = "ui",
  region = "794x98+137+909"
} --[[table: 0x7f41c6ec4590]]
07/15/23-10:26:05 DEBUG triggering refresh {
  mode = "ui",
  region = "808x152+130+867"
} --[[table: 0x7f41c3cf6110]]
07/15/23-10:26:07 DEBUG triggering refresh {
  mode = "ui",
  region = "1068x1887+0+0"
} --[[table: 0x7f41c556d250]]
07/15/23-10:26:07 DEBUG triggering refresh {
  mode = "fast",
  region = "160x109+908+0"
} --[[table: 0x7f41c4466c80]]
07/15/23-10:26:07 DEBUG triggering refresh {
  mode = "fast",
  region = "160x109+908+0"
} --[[table: 0x7f41c414eeb0]]
07/15/23-10:26:07 DEBUG triggering refresh {
  mode = "ui",
  region = "961x579+53+654"
} --[[table: 0x7f41cc10d8c0]]
07/15/23-10:26:09 DEBUG triggering refresh {
  mode = "flashui",
  region = "1068x1887+0+0"
} --[[table: 0x7f41c6e29818]]
07/15/23-10:26:09 DEBUG triggering refresh {
  mode = "flashui",
  region = "1068x716+0+0"
} --[[table: 0x7f41c6dca498]]
07/15/23-10:26:09 DEBUG triggering refresh {
  mode = "fast",
  region = "1054x59+7+135"
} --[[table: 0x7f41c702c9c8]]
07/15/23-10:26:09 DEBUG triggering refresh {
  mode = "flashui",
  region = "1068x1887+0+0"
} --[[table: 0x7f41c696c590]]
07/15/23-10:26:12 DEBUG triggering refresh {
  mode = "fast",
  region = "419x54+114+788"
} --[[table: 0x7f41c700cb80]]
07/15/23-10:26:12 DEBUG triggering refresh {
  mode = "ui",
  region = "1068x1887+0+0"
} --[[table: 0x7f41c701e338]]
07/15/23-10:26:14 DEBUG triggering refresh {
  mode = "flashui",
  region = "1068x716+0+0"
} --[[table: 0x7f41c6eed990]]
07/15/23-10:26:16 DEBUG triggering refresh {
  mode = "fast",
  region = "93x51+96+9"
} --[[table: 0x7f41c6f101d8]]
07/15/23-10:26:16 DEBUG triggering refresh {
  mode = "ui",
  region = "1068x716+0+0"
} --[[table: 0x7f41c6f474f8]]
```

</details>

One other thing, the wikipedia search function works normally, so koreader does have network connectivity, just not from the OPDS and Calibre Wireless functions.

**Comment by NiLuJe** [comment]
Verbose debug logs, please ;).

**Comment by zephyros-dev** [comment]
Here's the verbose output log for the reproduction steps with nightly build
[output.log](https://github.com/koreader/koreader/files/12060341/output.log)

**Comment by NiLuJe** [comment]
FWIW, I just realized I should be able to reproduce the environment by simply doing something like `env KO_MULTIUSER=1 ./kodev run`.

And, while, yeah, I can... I can't reproduce the issue ;o).

**Comment by zephyros-dev** [comment]
I think I found the issue. I haven't dig in the koreader source code, but there's seems to be a ping check to the default gateway in the OPDS function, and if it fails koreader will ask the user to turn on the network. Inside the WSL VM, this gateway is the host machine itself. By default, Windows firewall disallow receiving ICMP packets (https://github.com/microsoft/WSL/issues/4171#issuecomment-559961027), so the check will fail and ask the user to turn on the network even though it is working. The wikipedia functions probably works since it doesn't use the ping check.
I see 2 issues that should be dealt with here:
- The logs doesn't turn up anything regarding the ping check failure in the current version. I only managed to find out about this by reproducing the issue https://github.com/koreader/koreader/issues/10087#issuecomment-1636122720, and see that koreader is trying to ping the gateway and fail to do so.
- Should there be another way to check for network connectivity? I'm not sure how the Wikipedia functions does it (again, I haven't check the source code), but it works correctly in this case (It works normally when there's network connectivity, and asks for wifi prompt when I disable network on the host machine).
Tl;dr: Quick workaround: Allow Windows host machine to receive ICMPv4 packets in the firewall settings.

**Comment by Frenzie** [comment]
Network connectivity and being connected to the internet are two different things, that's why it works for Wikipedia. The scenario you describe is like how can this kid be his own grandfather. It's not strictly impossible but it doesn't make any sense.

I think the bigger question is if maybe the functions in question should always return true on SDL!emulator or something. The rationale for checking probably doesn't really exist in the first place.

**Comment by NiLuJe** [comment]
> I think the bigger question is if maybe the functions in question should always return true on SDL!emulator or something.

They do (since #10669), unless I missed one?

https://github.com/koreader/koreader/blob/3e43a21cf268e05f122a9493016c4c343229a92d/frontend/ui/network/manager.lua#L117-L130

**Comment by NiLuJe** [comment]
> unless I missed one?

Why, yes, yes I did forget `isOnline` ;o).

**Comment by NiLuJe** [comment]
Do not that *nothing* actually does a gateway ping (with the exception of the Network Info menu) in current master.

What `isOnline` does is trying to resolve `dns.msftncsi.com`, which,; ironically, is Windows's own connectivity check.

I'm not sure how that correlates to your WSL woes, though.

**Comment by Frenzie** [comment]
You might enjoy this picture of UTC in WSL vs time in Windows I took an hour ago.
![IMG20230715154455](https://github.com/koreader/koreader/assets/202757/badc47d0-2834-4d43-b8cd-06f23f9abd9b)

**Comment by zephyros-dev** [comment]
> Do note that _nothing_ actually does a gateway ping (with the exception of the Network Info menu) in current master.
> 
> What `isOnline` does is trying to resolve `dns.msftncsi.com`, which, ironically, is Windows's own connectivity check.
> 
> I'm not sure how that correlates to your WSL woes, though.

I'm pretty sure it does the gateway ping though, since I can reproduce the issue by blocking the ICMPv4 in the Windows Firewall settings again, and I found out about the ping by using the version with the issue in https://github.com/koreader/koreader/issues/10087#issuecomment-1636122720. The logs will print out the ping output when I do the OPDS step.

**Comment by Frenzie** [comment]
The correct version doesn't ping (isConnected), it just returns true.

What it doesn't return true for is this bit (isOnline):

https://github.com/koreader/koreader/blob/3bec20fc255a8e6a1ffa92ded361a451f367eeb9/frontend/ui/network/manager.lua#L398-L405

**Comment by alexandruchircu** [comment]
Can still reproduce this on KOReader 2023.10, WSL2 default Ubuntu dist fresh install, Windows 10, when trying to connect to OPDS.
